Respect Content-Type from internal headers instead of caching it in body

CYBAI opened the pull request:
#### cd292c4f555814e858cf649225cf2d26f036e58c
<pre>
Respect Content-Type from internal headers instead of caching it in body
<a href="https://bugs.webkit.org/show_bug.cgi?id=220535">https://bugs.webkit.org/show_bug.cgi?id=220535</a>

Reviewed by Youenn Fablet.

Previously, Content-Type is cached as `m_contentType` in
FetchBodyOwner. However, as the spec says, Content-Type should be
derived from the one in Headers. Thus, we&apos;d like to remove the cached
Content-Type and always retrieve it from internal headers instead.

* LayoutTests/http/wpt/fetch/fetch-as-blob-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.js: Removed.
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/cloned-any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.js:
(forEach.bodyContainerCreator.promise_test.async t):
(forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.promise_test.async t):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/serviceworker/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/window/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/worker/cache-match.https-expected.txt:
* Source/WebCore/Modules/fetch/FetchBody.cpp:
(WebCore::FetchBody::blob):
(WebCore::FetchBody::consumeOnceLoadingFinished):
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::resolveWithTypeAndData):
(WebCore::FetchBodyConsumer::resolve):
(WebCore::FetchBodyConsumer::takeAsBlob):
(WebCore::FetchBodyConsumer::clone):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
(WebCore::FetchBodyConsumer::setContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::blob):
(WebCore::FetchBodyOwner::cloneBody):
(WebCore::FetchBodyOwner::extractBody):
(WebCore::FetchBodyOwner::consumeOnceLoadingFinished):
(WebCore::FetchBodyOwner::formData):
(WebCore::FetchBodyOwner::updateContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.h:
(WebCore::FetchBodyOwner::contentType const):
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::m_signal):
(WebCore::FetchRequest::initializeWith):
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
(WebCore::FetchResponse::clone):
(WebCore::FetchResponse::setReceivedInternalResponse):

Canonical link: <a href="https://commits.webkit.org/264054@main">https://commits.webkit.org/264054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58c80f65289a23ad03d48a8ba6b7d6d6665e2202

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6825 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6702 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9722 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8210 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4268 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13748 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5990 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8386 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6474 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5296 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5875 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6247 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

CYBAI commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +138,36 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
+    auto result = FetchBody::extract(WTFMove(value), currentContentType);
     if (result.hasException())
         return result.releaseException();
+    m_headers->fastSet(HTTPHeaderName::ContentType, currentContentType);
```
@youennf @annevk Question here 🙋 

Previously, we're passing `m_contentType` as a reference to `FetchBody::extract`'s second argument and mutate it directly when necessary. However, as I change the `contentType` function to _not_ return the value as a reference, it wouldn't update to the m_headers and I need to do the `m_headers->fastSet` here. Would you think it's better to return a pair of `(Body, NewContentType)` from the extract function instead? I wonder it would be helpful to remind the next person who will use the function that the content type would potentially be updated.

annevk replied:
I'm not sure I understand the question. Doesn't `FetchBody::extract` essentially return a pair? One of them via an output parameter? This looks okay to me.

youennf replied:
Yeah, we could probably a later refactoring to have FetchBody::extract return all out parameters as return parameters. Probably not in this patch though.

CYBAI replied:
> I'm not sure I understand the question. Doesn't `FetchBody::extract` essentially return a pair? One of them via an output parameter? This looks okay to me.

oh, you are right 🙏  I didn't realize we can see it as an output parameter.

> Yeah, we could probably a later refactoring to have FetchBody::extract return all out parameters as return parameters. Probably not in this patch though.

yes, please let me know if you think we should refactor it and I will file an issue for it after this PR is merged 🙇

annevk commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +138,36 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
 ExceptionOr<void> FetchBodyOwner::extractBody(FetchBody::Init&& value)
 {
-    auto result = FetchBody::extract(WTFMove(value), m_contentType);
+    String currentContentType = contentType();
```
Nit: I suspect this works:
```suggestion
    auto currentContentType = contentType();
```

annevk reviewed (commented):
Nice, thanks for working on this!

youennf commented on Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
```diff
@@ -374,8 +374,8 @@ void FetchBodyConsumer::resolve(Ref<DeferredPromise>&& promise, const String& co
     case Type::Blob:
-        promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this, context](auto&) {
-            return takeAsBlob(context);
+        promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this, context, contentType](auto&) {
```
&contentType.
Also we can probably removing capturing context.
```
promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this, &contentType](auto& context) {
```

youennf commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -129,7 +129,7 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
     if (isBodyNullOrOpaque()) {
         auto* context = promise->scriptExecutionContext();
         promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this, context](auto&) {
-            return Blob::create(context, Vector<uint8_t> { }, Blob::normalizedContentType(extractMIMETypeFromMediaType(m_contentType)));
```
We can probably do some additional refactoring here: remove the context variable and get it from the function.
```
promise->resolveCallbackValueWithNewlyCreated<IDLInterface<Blob>>([this](auto& context) {
```

CYBAI replied:
thanks for the tips! I will fix the capturing context in upcoming commits.

youennf commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -285,7 +274,8 @@ void FetchBodyOwner::blobLoadingSucceeded()
 
-    m_body->loadingSucceeded(contentType());
+    String contentType = m_headers->fastGet(HTTPHeaderName::ContentType);
+    m_body->loadingSucceeded(contentType);
```
We probably do not need this change.

CYBAI replied:
ah. good catch! I will change it back 🙏

CYBAI commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +137,42 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
+
+    // Initialize the Content-Type header if it was empty.
+    if (wasEmptyContentType)
+        m_headers->fastSet(HTTPHeaderName::ContentType, currentContentType);
```
I end up realizing that we should only update the content type to headers when it was empty before running `FetchBody::extract`. If any existing content-type already exists in headers, we should keep it as is.

annevk replied:
Note that empty and null are not the same thing. We probably want to call this contentTypeIsMissing or some such as that's different from it being the empty string.

CYBAI replied:
ah, good point! Thanks for catching this. Let me rename the variable.

CYBAI commented on LayoutTests/http/wpt/fetch/fetch-as-blob.js:
```diff
@@ -66,7 +66,7 @@ promise_test(function(t) {
     return req.blob()
       .then(function(blob) {
-          assert_equals(blob.type, 'text/plain');
+          assert_equals(blob.type, 'text/html');
```
I just updated the test here to confirm there's no other issues. But, I've sent a PR to add this test to upstream wpt so I will import that upstream tests and remove this one when the upstream PR is merged.

annevk replied:
I've reviewed and merged https://github.com/web-platform-tests/wpt/pull/39489.

The WebKit changes look reasonable to me, but @youennf will need to do a final review.

CYBAI replied:
Thank you so much for your help to merge the WPT PR 🙏 
I will import them and update this PR later 🙇

CYBAI commented on LayoutTests/http/wpt/fetch/fetch-as-blob.js:
```diff
@@ -1,87 +0,0 @@
-  importScripts('/resources/testharness.js');
-}
-
-promise_test(function(t) {
```
These tests in this file are already merged into the `fetch/api/body/mime-type.any.js` WPT tests so this one is removed in this PR.

youennf commented on Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
```diff
@@ -385,6 +383,7 @@ void FetchBodyConsumer::resolve(Ref<DeferredPromise>&& promise, const String& co
         promise->resolve<IDLDOMString>(takeAsText());
         return;
     case FetchBodyConsumer::Type::FormData: {
+        auto* context = promise->scriptExecutionContext();
```
It is only used once, let's call packageFormData with `promise->scriptExecutionContext()` directly

youennf commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +137,42 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
     m_loadingError = owner.m_loadingError;
-
-    m_contentType = owner.m_contentType;
+    m_headers = FetchHeaders::create(owner.m_headers.get());
```
Why is this needed?
We are only supposed to clone the body here, callers of cloneBody should clone the headers themselves?
This might be to handle FetchResponse::clone, in which case we might actually want to change FetchResponse::clone.

CYBAI replied:
no, thanks for catching this! I will move this logic to `FetchResponse::clone` and leave this function only take care of the clone of body 🙏 (we don't need to clone headers in `FetchRequest::clone` because it already pass the headers to `FetchRequest::create`)

youennf commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +137,42 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
 {
-    auto result = FetchBody::extract(WTFMove(value), m_contentType);
+    auto currentContentType = contentType();
+    auto wasEmptyContentType = currentContentType.isNull();
```
wasEmptyContentType is not a great name, the isNull check means there was no content-type header, it does not mean there was a content-type header with an empty value.

CYBAI replied:
you're right. please let me rename it to `contentTypeNotExist` 🙏

youennf reviewed (commented):
This looks almost ready to me and is a nice simplification.
Some questions/suggestions below.

youennf commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -138,47 +137,41 @@ void FetchBodyOwner::blob(Ref<DeferredPromise>&& promise)
 {
-    auto result = FetchBody::extract(WTFMove(value), m_contentType);
+    auto currentContentType = contentType();
+    auto contentTypeNotExist = currentContentType.isNull();
```
```suggestion
    bool isContentTypeSet = !currentContentType.isNull();
```

youennf replied:
And then need to change `contentTypeNotExist ` by `!isContentTypeSet` below

CYBAI replied:
thanks for the suggestion 🙏

CYBAI committed:
Respect Content-Type from internal headers instead of caching it in body
https://bugs.webkit.org/show_bug.cgi?id=220535

Reviewed by Youenn Fablet.

Previously, Content-Type is cached as `m_contentType` in
FetchBodyOwner. However, as the spec says, Content-Type should be
derived from the one in Headers. Thus, we'd like to remove the cached
Content-Type and always retrieve it from internal headers instead.

* LayoutTests/http/wpt/fetch/fetch-as-blob-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker-expected.txt: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob-worker.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.html: Removed.
* LayoutTests/http/wpt/fetch/fetch-as-blob.js: Removed.
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/cloned-any.js: Added.
(promise_test.async t):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.js:
(forEach.bodyContainerCreator.promise_test.async t):
(forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.promise_test.async t):
(new.Response.new.Blob.string_appeared_here.forEach.bodyContainerCreator.assert_equals):
(new.Response.new.Blob):
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/mime-type.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/body/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/serviceworker/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/window/cache-match.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/worker/cache-match.https-expected.txt:
* Source/WebCore/Modules/fetch/FetchBody.cpp:
(WebCore::FetchBody::blob):
(WebCore::FetchBody::consumeOnceLoadingFinished):
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
(WebCore::resolveWithTypeAndData):
(WebCore::FetchBodyConsumer::resolve):
(WebCore::FetchBodyConsumer::takeAsBlob):
(WebCore::FetchBodyConsumer::clone):
* Source/WebCore/Modules/fetch/FetchBodyConsumer.h:
(WebCore::FetchBodyConsumer::setContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::blob):
(WebCore::FetchBodyOwner::cloneBody):
(WebCore::FetchBodyOwner::extractBody):
(WebCore::FetchBodyOwner::consumeOnceLoadingFinished):
(WebCore::FetchBodyOwner::formData):
(WebCore::FetchBodyOwner::updateContentType): Deleted.
* Source/WebCore/Modules/fetch/FetchBodyOwner.h:
(WebCore::FetchBodyOwner::contentType const):
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::m_signal):
(WebCore::FetchRequest::initializeWith):
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
(WebCore::FetchResponse::clone):
(WebCore::FetchResponse::setReceivedInternalResponse):

Canonical link: https://commits.webkit.org/264054@main
